Bulk ingest: single FOFN  + Quickstart + Beta [VS-962] [VS-984] [VS-982]

mcovarr opened the pull request:
This ticket set out to do VS-962 (one FOFN to rule them all) only, but along the way I found I also needed to do VS-984 (use Bulk Ingest from integration test) as the interface to `GvsAssignIds` had changed. After making these changes I realized I had basically done VS-982 as well (use Bulk Ingest in Beta WDL) due to the beta WDL picking up the changes to the unified WDL both it and Quickstart call.

- [Successful integration run](https://job-manager.dsde-prod.broadinstitute.org/jobs/ab07dffd-a2ea-4b69-9c4c-eec0019e5b3b)
- [Pending beta run](https://job-manager.dsde-prod.broadinstitute.org/jobs/37f6be29-5aae-42c4-a86b-c6d00c3caec5)

koncheto-broad commented on scripts/variantstore/wdl/GvsBulkIngestGenomes.wdl:
```diff
@@ -79,26 +80,24 @@ workflow GvsBulkIngestGenomes {
             project_id = project_id,
-            external_sample_names = read_lines(PrepareBulkImport.sampleFOFN),
-            samples_are_controls = false
+            bulk_import_tsv = PrepareBulkImport.output_tsv,
```
I imagined a step here between PrepareBulkImport and GvsAssignIds where we can keep our original interface so make development easier.  See my comment above in AssignIds

koncheto-broad reviewed (commented):
I think I interpreted the general idea of "our only ingest will be bulk ingest" as more of a policy on what functionality we'll support for users (and us when doing AoU).  For dev purposes, changing the inputs for AssignIds and ImportGenomes to a tsv feels like it puts an extra, non-trivial step in the way.  Is it strictly necessary?  Can't we let the bulk ingest portion spit out a tsv so we know its output is in sync then just split that file and pass it into the individual pipeline pieces like before?

gbggrant commented on scripts/variantstore/wdl/GvsImportGenomes.wdl:
```diff
@@ -254,39 +218,47 @@ task LoadData {
 
     ## now we want to create a sub list of these samples (without the ones that have already been loaded)
-    curl https://raw.githubusercontent.com/broadinstitute/gatk/ah_var_store/scripts/variantstore/wdl/extract/curate_input_array_files.py -o curate_input_array_files.py
+    curl --location --remote-name https://raw.githubusercontent.com/broadinstitute/gatk/~{branch_name}/scripts/variantstore/wdl/extract/curate_bulk_import_tsv.py
```
Wouldn't it be better to just put this python script into the docker itself, rather than using git to pull it from the repo? I think we should limit passing branch names as much as possible in production code.

mcovarr replied:
That would require making an additional Variants image based on the GATK image. Not out of the question, but I don't think that's something for this PR.

mcovarr replied:
Options to look at:

- Javaizing this and putting it into the GATK image
- Figure out how to get Python script onto GATK image
- Passing it in from a precursor task

gbggrant commented on scripts/variantstore/wdl/GvsBulkIngestGenomes.wdl:
```diff
@@ -37,6 +37,7 @@ workflow GvsBulkIngestGenomes {
         Int? load_data_preemptible_override
         Int? load_data_maxretries_override
         # End GvsImportGenomes
+        String branch_name = "ah_var_store"
```
Unrelated entirely to your work - but call_set_identifier (above) is unused. I'll give you a nickel if you remove it.

mcovarr replied:
Yeah I found there's some other unused stuff as well; I'll make an additional cleanup pass.

mcovarr commented on scripts/variantstore/wdl/GvsAssignIds.wdl:
```diff
@@ -9,7 +9,7 @@ workflow GvsAssignIds {
     String project_id
 
-    Array[String] external_sample_names
+    File bulk_import_tsv
```
I was unaware that there was a desire to run any of the subworkflows called by `GvsBulkIngest` individually during dev. We can certainly make changes to support this but it would be helpful to capture the requirements first.

mcovarr commented:
Re-runs with PR feedback incorporated:

- [Integration](https://app.terra.bio/#workspaces/gvs-dev/mlc%20GVS%20Quickstart%203%20samples/job_history/e8b6077d-a90a-4cc2-be0d-0a08cb98280a)
- [Beta](https://app.terra.bio/#workspaces/gvs-dev/mlc%20GVS%20Quickstart%203%20samples/job_history/52a3c02e-485b-4320-bb21-07931ecbe7dd)

rsasch reviewed (changes requested):
- is there a reason to keep GvsPrepareBulkImport.wdl and GvsBulkIngestGenomes.wdl as two separate WDLs?
- can we emove GvsUnified.wdl and use GvsJointVariantCalling.wdl to call the steps?
- fix in get_columns_for_import.py (column -> col from https://github.com/broadinstitute/gatk/pull/8379/files?w=0)

mcovarr commented:
> is there a reason to keep GvsPrepareBulkImport.wdl and GvsBulkIngestGenomes.wdl as two separate WDLs?

Not AFAICT, addressed in [this commit](https://github.com/broadinstitute/gatk/pull/8404/commits/2a0bf3bb2bba91f98e5b905b150d5888f192e84e).

> can we emove GvsUnified.wdl and use GvsJointVariantCalling.wdl to call the steps?

I'm not sure how that would work. `GvsJointVariantCalling.wdl` intentionally hardcodes several parameters to `GvsUnified` that `GvsQuickstartIntegration.wdl` needs to override:

- `gatk_override`
- `extract_table_prefix`
- `interval_list`
- `use_VQSR_lite`
- `extract_do_not_filter_override`
- `filter_set_name`

Happy to discuss further if I'm missing something here.

> fix in get_columns_for_import.py (column -> col from https://github.com/broadinstitute/gatk/pull/8379/files?w=0)

Thanks, added [here](https://github.com/broadinstitute/gatk/pull/8404/commits/d628a8ef34c8dbb3c0efaa51cd93bb28e2a18058). Do we know why there isn't an existing test that would have found this?

rsasch commented:
> * `gatk_override`

this will need to be a parameter (but can be optional and have a default) in order to have our integration tests run `GvsJointVariantCalling.wdl` anyway

> * `interval_list`

this will need to be a parameter (but can be optional and have a default) in order to have our integration tests run `GvsJointVariantCalling.wdl` on exomes

> * `use_VQSR_lite`
> * `extract_do_not_filter_override`

these two will need to be a parameter (but can be optional and have a default) in order to have our integration tests run `GvsJointVariantCalling.wdl` with VQSR classic anyway

> * `filter_set_name`
> * `extract_table_prefix`

these two can just default to the `call_set_identifier`  with weird characters parsed out

mcovarr commented:
>> interval_list

> this will need to be a parameter (but can be optional and have a default) in order to have our integration tests run GvsJointVariantCalling.wdl on exomes

Genomes too, the integration test specifies a 20/X/Y interval list.

>> filter_set_name
>> extract_table_prefix
these two can just default to the call_set_identifier with weird characters parsed out


@RoriCremer can correct me if I'm wrong, but I thought the raison d'être of the beta WDL was specifically to hardcode away as many parameters as possible (even optional ones with defaults) to present a simplified interface for non-expert users. I agree we'll probably have to allow some additional parameters for testability (`gatk_override` at a minimum), but do we really want all of these?

mcovarr commented:
😭 https://github.com/broadinstitute/gatk/actions/runs/5547450688/jobs/10131043668?pr=8404
```
org.broadinstitute.hellbender.tools.walkers.rnaseq.SplitNCigarReadsIntegrationTest > testLargeFileThatForcesSnappyUsage FAILED
    htsjdk.samtools.SAMException: Exception when processing alignment for BAM index 20FUKAAXX100202:4:48:16031:99011 2/2 101b aligned to 21:10185002-10185087.
        at htsjdk.samtools.BAMFileWriter.writeAlignment(BAMFileWriter.java:141)
        at htsjdk.samtools.SAMFileWriterImpl.close(SAMFileWriterImpl.java:212)
        at htsjdk.samtools.AsyncSAMFileWriter.synchronouslyClose(AsyncSAMFileWriter.java:38)
        at htsjdk.samtools.util.AbstractAsyncWriter.close(AbstractAsyncWriter.java:89)
        at org.broadinstitute.hellbender.utils.read.SAMFileGATKReadWriter.close(SAMFileGATKReadWriter.java:26)
        at org.broadinstitute.hellbender.tools.walkers.rnaseq.SplitNCigarReads.closeTool(SplitNCigarReads.java:193)
        at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:1064)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:140)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:[192](https://github.com/broadinstitute/gatk/actions/runs/5547450688/jobs/10131043668#step:12:192))
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:211)
        at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
        at org.broadinstitute.hellbender.Main.instanceMain(Main.java:146)
        at org.broadinstitute.hellbender.Main.instanceMain(Main.java:187)
        at org.broadinstitute.hellbender.CommandLineProgramTest.runCommandLine(CommandLineProgramTest.java:27)
        at org.broadinstitute.hellbender.testutils.CommandLineProgramTester.runCommandLine(CommandLineProgramTester.java:111)
        at org.broadinstitute.hellbender.tools.walkers.rnaseq.SplitNCigarReadsIntegrationTest.testLargeFileThatForcesSnappyUsage(SplitNCigarReadsIntegrationTest.java:85)
        Caused by:
        htsjdk.samtools.util.RuntimeIOException: Write error; BinaryCodec in writemode; streamed file (filename not available)
            at htsjdk.samtools.util.BinaryCodec.writeBytes(BinaryCodec.java:222)
            at htsjdk.samtools.util.BlockCompressedOutputStream.writeGzipBlock(BlockCompressedOutputStream.java:444)
            at htsjdk.samtools.util.BlockCompressedOutputStream.deflateBlock(BlockCompressedOutputStream.java:408)
            at htsjdk.samtools.util.BlockCompressedOutputStream.write(BlockCompressedOutputStream.java:301)
            at htsjdk.samtools.util.BinaryCodec.writeBytes(BinaryCodec.java:220)
            at htsjdk.samtools.util.BinaryCodec.writeBytes(BinaryCodec.java:212)
            at htsjdk.samtools.BAMRecordCodec.encode(BAMRecordCodec.java:168)
            at htsjdk.samtools.BAMFileWriter.writeAlignment(BAMFileWriter.java:134)
            ... 15 more
            Caused by:
            java.io.IOException: No space left on device
                at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
                at sun.nio.ch.FileDispatcherImpl.write(FileDispatcherImpl.java:60)
                at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
                at sun.nio.ch.IOUtil.write(IOUtil.java:65)
                at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:[211](https://github.com/broadinstitute/gatk/actions/runs/5547450688/jobs/10131043668#step:12:211))
                at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
                at java.nio.channels.Channels.writeFully(Channels.java:101)
                at java.nio.channels.Channels.access$000(Channels.java:61)
                at java.nio.channels.Channels$1.write(Channels.java:174)
                at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
                at java.io.BufferedOutputStream.write(BufferedOutputStream.java:126)
                at htsjdk.samtools.util.BinaryCodec.writeBytes(BinaryCodec.java:220)
                ... 22 more
```

gbggrant reviewed (approved):
Looks good.
Approved, provided discussion of removal of GvsUnified is resolved.

mcovarr commented:
Integration run in progress with `GvsUnified` removed [here](https://job-manager.dsde-prod.broadinstitute.org/jobs/52748da4-46e6-4ae4-97a4-f91ca6517e34)

rsasch commented on scripts/variantstore/wdl/GvsAssignIds.wdl:
```diff
@@ -128,8 +128,8 @@ task AssignIds {
   command <<<
-    set -e
-    set -x
+    PS4='\D{+%F %T} \w $ '
```
nit: might want to add the comment `# Prepend date, time and pwd to xtrace log entries.` like you did elsewhere for clarity

rsasch commented on scripts/variantstore/wdl/GvsBulkIngestGenomes.wdl:
```diff
@@ -69,36 +67,40 @@ workflow GvsBulkIngestGenomes {
+    call GenerateImportFofnFromDataTable {
         input:
-            project_id = GetWorkspaceName.workspace_namespace,
+            google_project_id = GetWorkspaceName.workspace_namespace,
```
nit: maybe a better name would be `terra_google_project_id` or `workspace_google_project`?

rsasch commented on scripts/variantstore/wdl/GvsImportGenomes.wdl:
```diff
@@ -428,6 +415,7 @@ task GetUningestedSampleIds {
   String temp_table="~{dataset_name}.sample_names_to_load"
 
   command <<<
+    PS4='\D{+%F %T} \w $ '
```
see above

rsasch commented on scripts/variantstore/wdl/GvsJointVariantCalling.wdl:
```diff
@@ -1,27 +1,47 @@
         Boolean use_classic_VQSR = true
         # Beta users have accounts with tighter quotas, and we must work around that
         Boolean tighter_gcp_quotas = true
+        # TODO should these all be exposed in this WDL?
```
nit: do we need this TODO?

mcovarr replied:
nope

rsasch commented on scripts/variantstore/wdl/GvsJointVariantCalling.wdl:
```diff
@@ -1,27 +1,47 @@
+        String? sample_set_name ## NOTE: currently we only allow the loading of one sample set at a time
+
+        # This is the most updated snapshot of the code as of July 2, 2023
+        File gatk_override = "gs://gvs_quickstart_storage/jars/gatk-package-4.2.0.0-724-gf478fb2-SNAPSHOT-local.jar"
```
do we want to use the more recent gs://gvs_quickstart_storage/jars/gatk-package-4.2.0.0-727-g950da0b-SNAPSHOT-local.jar (July 12, 2023)?

mcovarr replied:
mmm actually I'm going to need to bake a fresh one based that includes everything in the July 12 one PLUS the curation script that now stows away in GATK

rsasch replied:
🤦🏻‍♀️ ugh, sorry

rsasch reviewed (approved):
Thanks for all the changes!  Just one concern: please check that the GATK docker you're using is the most up to date; it looks like one with the label `varstore_2023_07_12` uploaded after `varstore_2023_07_12_3ba8f86`.

koncheto-broad reviewed (approved):
Thanks for addressing the issues with the individual WDLs remaining separately runnable! LGTM
